versions/schematic-2.md

```diff
@@ -0,0 +1,235 @@
+<a name="schematicMetadata"></a>Metadata | [Metadata Object](#metadataObject) as `NBT compound` | **Optional.** Provides metadata about the schematic. This shall be saved as [NBT](#formFile) tag compound.
+<a name="schematicWidth"></a>Width | `NBT int` | **Required.** Specifies the width (the amount of blocks along the X-axis) of the schematic, must be interpreted as unsigned. This shall be saved as [NBT](#formFile) tag int.
+<a name="schematicHeight"></a>Height | `NBT int` | **Required.** Specifies the height (the amount of blocks along the Y-axis) of the schematic, must be interpreted as unsigned. This shall be saved as [NBT](#formFile) tag int.
+<a name="schematicLength"></a>Length | `NBT int` | **Required.** Specifies the length (the amount of blocks along the Z-axis) of the schematic, must be interpreted as unsigned. This shall be saved as [NBT](#formFile) tag int.
+<a name="schematicOffset"></a>Offset | `NBT int array`[3] | **Optional.** Specifies the relative offset of the schematic. This value is relative to the most minimal point in the schematics area. If present, this offset shall be added to all blocks, tiles and entities when pasting it into the world (may be negative). This shall be saved as [NBT](#formFile) tag int array.
+<a name="schematicPaletteMax"></a>PaletteMax | `NBT byte` | **Required.** Specifies the size of the block palette index in number of bytes needed for the maximum palette index. Implementations must use this to determine, which index format will be needed. The palette index may fit within a datatype smaller than a 32-bit integer or even a longer datatype like 64-bit integer. This shall be saved as [NBT](#formFile) tag byte.
+<a name="schematicPalette"></a>Palette | `NBT list` of [Palette Object](#paletteObject) as `NBT compound` | **Required.** Specifies the block palette. This is a mapping of block states to indices which are local to this schematic. The array element indices are used to reference the block states from within the [BlockData array](#schematicBlockData). The maximum index cannot be greater than [`2^(PaletteMax * 8) - 1`](#schematicPaletteMax).  This shall be saved as [NBT](#formFile) tag list of [NBT](#formFile) tag string.
+<a name="schematicBlockData"></a>BlockData | `nbt int array` or `nbt long array` | **Required.** Specifies the main storage array which contains `Width * Height * Length` entries. Each entry is specified as a `int` or `long`, depending on the given [PaletteMax](#schematicPaletteMax) (1 to 4 bytes = int; 5 to 8 bytes = long). This index refers to the array element index within the [Palette](#schematicPalette). The entries are indexed by `x + z * Width + y * Width * Length`. This shall be saved as [NBT](#formFile) tag int array or [NBT](#formFile) tag long array.
```

Comment:
Since we agreed now to use shorts because of the Java array index limitation, forget about longs...
Same about the "running out of indices" problem.

I just used int instead of varint, because NBT doesn't have varints and i don't know any language, which supports varints natively.

Reply:
Using varints over ints is a really important space optimization, ideally your palette should be sorted such that the most used blocks are mapped to indices smaller than 128 which means they only need a single byte to store. Using ints instead just increased the filesize by a factor of 4. And they're seriously not that hard to work with: https://github.com/SpongePowered/SpongeCommon/blob/stable-7/src/main/java/org/spongepowered/common/data/persistence/SchematicTranslator.java#L148-L175

Reply:
No language supports varints natively because they're a storage concept. A lot of formats and protocols use them for space optimization, including minecraft.